# ACCY 577 Final Project  

## Team Members
Please list your team members in this cell.

----



## Overview 


After several years of operation, the Lending Club&trade; wants to understand how the individual components (i.e., features) of a loan, such as the amount loaned, the term length, or the interest rate, affect the profitability of a making a specific loan. In particular, they're interested in understanding how likely a loan is to be repaid. In this project, you will use the provided loans dataset to help the Lending Club&trade; figure out which loans are most profitable.


## Criteria

You will work in groups to analyze the data to make recommendations based on the `lending_club_2007_2011_6_states.csv` dataset. 

You will complete three tasks for this group project:
1. A group report in the form of a Jupyter notebook,
2. A video presentation where your group will present your results, and 
3. Peer evaluation of the contributions of each member of your group.

Your final group report will be a single Jupyter notebook that will integrate Markdown, Python code, and the results from your code, such as data visualizations. Markdown cells should be used to explain any decisions you make regarding the data, to discuss any plots or visualizations generated in your notebook, and the results of your analysis. As a general guideline, the content should be written in a way that a fellow classmate (or an arbitrary data scientist/analyst) should be able to read your report and understand the results, implications, and processes that you followed to achieve your result.

All group members should present in the video presentation. You can use presentation software such as MS PowerPoint. The presentations should cover the steps in your analytics process and highlight your results. You don't need to explain python code in the presentation. Focus on your analysis method and results. The presentation should take between eight to twelve minutes.

### Rubric
  - Notebook Report (160 points).
  - Video presentation (100 points).
  - Peer assessment for your group-mates (20 points). You will get this 20 points as long as you submit peer evaluation before deadline.
  - Inidividual grade will be adjusted based on your peer evaluations by your group members.
  
### Submissions
- Each Group:
 - Report in .ipynb format
 - Report in .html format
 - Presentation link
- Each individual:
 - Peer evaluation

### General

Your report should 
  - Use proper markdown.  
  - Include all of the code used for your analysis.
  - Plots should be properly labeled (e.g., axis labels and titles).
  - Use a consistent style.
  
Your presentation should
  - Use clear, concise slides.
  - Support your points with facts (e.g., statistics, visualizations).
  - Avoid any coding details.
  - Be creative.
-----

## Get Started

In [1]:
%matplotlib inline

import numpy as np
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt

#display all dataframe columns in df.head()
pd.options.display.max_columns = None
#display long string in datafame
pd.options.display.max_colwidth = 300

#filter out warning messages
import warnings
warnings.filterwarnings('ignore')

### Business Understanding

#### Lending Club

LendingClub is an American peer-to-peer lending company, headquartered in San Francisco, California. It is the world's largest peer-to-peer lending platform.

LendingClub enables borrowers to create unsecured personal loans between \\$1,000 and \\$40,000. Investors can search and browse the loan listings on LendingClub website and select loans that they want to invest in based on the information supplied about the borrower, amount of loan, loan grade, and loan purpose. Investors make money from interest. LendingClub makes money by charging borrowers an origination fee and investors a service fee.

For more information about the company please check out the Wikipedia article about the [LendingClub](https://en.wikipedia.org/wiki/LendingClub).


### Load Data and Data Dictionary

In [2]:
data_dict = pd.read_csv('data_dictionary.csv')
data_dict

ColumnName  \
0                acc_now_delinq   
1                    addr_state   
2                    annual_inc   
3              application_type   
4      chargeoff_within_12_mths   
5       collection_recovery_fee   
6    collections_12_mths_ex_med   
7          debt_settlement_flag   
8     debt_settlement_flag_date   
9                   delinq_2yrs   
10                  delinq_amnt   
11          disbursement_method   
12                          dti   
13             earliest_cr_line   
14                   emp_length   
15                    emp_title   
16                  funded_amnt   
17              funded_amnt_inv   
18                        grade   
19                hardship_flag   
20               home_ownership   
21          initial_list_status   
22               inq_last_6mths   
23                  installment   
24                     int_rate   
25                      issue_d   
26           last_credit_pull_d   
27              last_pymnt_amnt   
28                 last_pymnt_d   
29                    loan_amnt   
30                  loan_status   
31       mths_since_last_delinq   
32  mths_since_last_major_derog   
33       mths_since_last_record   
34                 next_pymnt_d   
35                     open_acc   
36                    out_prncp   
37                out_prncp_inv   
38                  policy_code   
39                      pub_rec   
40         pub_rec_bankruptcies   
41                      purpose   
42                   pymnt_plan   
43                   recoveries   
44                    revol_bal   
45                   revol_util   
46                    sub_grade   
47                    tax_liens   
48                         term   
49                        title   
50                    total_acc   
51                  total_pymnt   
52              total_pymnt_inv   
53                total_rec_int   
54           total_rec_late_fee   
55              total_rec_prncp   
56          verification_status   
57                     zip_code   

                                                                                                                                                                                                 Description  
0                                                                                                                                            The number of accounts on which the borrower is now delinquent.  
1                                                                                                                                                 The state provided by the borrower in the loan application  
2                                                                                                                              The self-reported annual income provided by the borrower during registration.  
3                                                                                                       Indicates whether the loan is an individual application or a joint application with two co-borrowers  
4                                                                                                                                                                     Number of charge-offs within 12 months  
5                                                                                                                                                                             post charge off collection fee  
6                                                                                                                                           Number of collections in 12 months excluding medical collections  
7                                                                                                         Flags whether or not the borrower, who has charged-off, is working with a debt-settlement company.  
8                                                                                           

In [3]:
loan_df = pd.read_csv('lending_club_2007_2011_6_states.csv')
loan_df.sample(5)

loan_amnt  funded_amnt  funded_amnt_inv       term  int_rate  \
6862       15000        15000     15000.000000  36 months     11.49   
10396       5500         5500      5475.000000  36 months      6.92   
7587       18000        17125     17099.710661  60 months     13.99   
2478        4000         4000      4000.000000  36 months      6.62   
5800        6000         6000      5750.000000  36 months      7.49   

       installment grade sub_grade                 emp_title emp_length  \
6862        494.57     B        B4           yellow cab taxi    6 years   
10396       169.63     A        A3                   ET Horn    3 years   
7587        398.39     C        C3   brewster central school  10+ years   
2478        122.82     A        A2                       NaN        NaN   
5800        186.61     A        A4  Winston Retail Solutions    4 years   

      home_ownership  annual_inc verification_status   issue_d loan_status  \
6862            RENT     36000.0        Not Verified  Jun-2011  Fully Paid   
10396           RENT     45000.0        Not Verified  Jan-2011  Fully Paid   
7587        MORTGAGE     40000.0            Verified  May-2011  Fully Paid   
2478        MORTGAGE     44000.0        Not Verified  Oct-2011  Fully Paid   
5800            RENT     76000.0        Not Verified  Jul-2011  Fully Paid   

      pymnt_plan             purpose                      title zip_code  \
6862           n  debt_consolidation                  prius2008    945xx   
10396          n  debt_consolidation  Credit Card consolidation    906xx   
7587           n         credit_card           Credit Card Loan    125xx   
2478           n      small_business        Small Business Loan    921xx   
5800           n  debt_consolidation   Cap One and Neimans Debt    941xx   

      addr_state    dti  delinq_2yrs earliest_cr_line  inq_last_6mths  \
6862          CA  11.93          0.0         Dec-2006             3.0   
10396         CA  16.83          0.0         Nov-2002             0.0   
7587          NY  21.12          0.0         Apr-1976             4.0   
2478          CA   1.53          0.0         May-1973             0.0   
5800          CA  17.12          0.0         Jan-1994             0.0   

       mths_since_last_delinq  mths_since_last_record  open_acc  pub_rec  \
6862                      NaN                     NaN       4.0      0.0   
10396                     NaN                     NaN      10.0      0.0   
7587                      NaN                     NaN      16.0      0.0   
2478                      NaN                     NaN       4.0      0.0   
5800                      NaN                     NaN       6.0      0.0   

       revol_bal  revol_util  total_acc initial_list_status  out_prncp  \
6862         289        12.6        8.0                   f        0.0   
10396       4241        30.7       20.0                   f        0.0   
7587       20287        47.2       32.0                   f        0.0   
2478        2709        15.9       10.0                   f        0.0   
5800       24493        68.0       22.0                   f        0.0   

       out_prncp_inv   total_pymnt  total_pymnt_inv  total_rec_prncp  \
6862             0.0  17448.493034         17448.49          15000.0   
10396            0.0   6007.201116          5979.89           5500.0   
7587             0.0  23907.713117         23872.23          17125.0   
2478             0.0   4421.305870          4421.31           4000.0   
5800             0.0   6717.950109          6438.04           6000.0   

       total_rec_int  total_rec_late_fee  recoveries  collection_recovery_fee  \
6862         2423.76               24.73         0.0                      0.0   
10396         507.20                0.00         0.0                      0.0   
7587         6782.71                0.00         0.0                      0.0   
2478          421.31                0.00         0.0                      0.0   
5800          717.95    

In [4]:
loan_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 19908 entries, 0 to 19907
Data columns (total 58 columns):
loan_amnt                      19908 non-null int64
funded_amnt                    19908 non-null int64
funded_amnt_inv                19908 non-null float64
term                           19908 non-null object
int_rate                       19908 non-null float64
installment                    19908 non-null float64
grade                          19908 non-null object
sub_grade                      19908 non-null object
emp_title                      18723 non-null object
emp_length                     19409 non-null object
home_ownership                 19908 non-null object
annual_inc                     19908 non-null float64
verification_status            19908 non-null object
issue_d                        19908 non-null object
loan_status                    19908 non-null object
pymnt_plan                     19908 non-null object
purpose                        19908 non-

In [5]:
loan_df.describe()

loan_amnt   funded_amnt  funded_amnt_inv      int_rate  \
count  19908.000000  19908.000000     19908.000000  19908.000000   
mean   11353.846444  11065.763763     10500.929748     12.089717   
std     7463.700492   7176.276661      7106.229650      3.698287   
min      500.000000    500.000000         0.000000      5.420000   
25%     5750.000000   5600.000000      5000.000000      9.450000   
50%    10000.000000  10000.000000      9000.000000     11.860000   
75%    15000.000000  15000.000000     14500.000000     14.610000   
max    35000.000000  35000.000000     35000.000000     24.400000   

        installment    annual_inc           dti   delinq_2yrs  inq_last_6mths  \
count  19908.000000  1.990800e+04  19908.000000  19908.000000    19908.000000   
mean     330.614254  7.107345e+04     13.008619      0.143962        0.829466   
std      210.557434  6.980565e+04      6.663658      0.489576        1.044978   
min       15.690000  4.000000e+03      0.000000      0.000000        0.000000   
25%      171.287500  4.200000e+04      7.840000      0.000000        0.000000   
50%      285.780000  6.000000e+04     13.075000      0.000000        0.000000   
75%      440.815000  8.500000e+04     18.200000      0.000000        1.000000   
max     1302.690000  6.000000e+06     29.990000     11.000000        8.000000   

       mths_since_last_delinq  mths_since_last_record      open_acc  \
count             6998.000000             1209.000000  19908.000000   
mean                35.841097               69.354839      9.278782   
std                 21.599170               44.520279      4.414903   
min                  0.000000                0.000000      2.000000   
25%                 19.000000                0.000000      6.000000   
50%                 34.000000               90.000000      9.000000   
75%                 51.000000              104.000000     12.000000   
max                106.000000              129.000000     44.000000   

            pub_rec      revol_bal    revol_util     total_acc  out_prncp  \
count  19908.000000   19908.000000  19889.000000  19908.000000    19908.0   
mean       0.046715   13363.994826     49.424966     21.527627        0.0   
std        0.217827   15943.303849     28.198395     11.269006        0.0   
min        0.000000       0.000000      0.000000      2.000000        0.0   
25%        0.000000    3770.500000     26.500000     13.000000        0.0   
50%        0.000000    8876.500000     50.300000     20.000000        0.0   
75%        0.000000   16952.250000     72.800000     28.000000        0.0   
max        3.000000  148829.000000     99.900000     90.000000        0.0   

       out_prncp_inv   total_pymnt  total_pymnt_inv  total_rec_prncp  \
count        19908.0  19908.000000     19908.000000     19908.000000   
mean             0.0  12286.852391     11690.155107      9913.519990   
std              0.0   9098.847567      8984.722396      7126.424699   
min              0.0      0.000000         0.000000         0.000000   
25%              0.0   5670.592545      5239.857500      4800.000000   
50%              0.0  10042.735817      9427.845000      8000.000000   
75%              0.0  16681.688450     15983.440000     14000.000000   
max              0.0  58480.139915     58438.370000     35000.020000   

       total_rec_int  total_rec_late_fee    recoveries  \
count   19908.000000        19908.000000  19908.000000   
mean     2277.120129            1.489973     94.722382   
std      2583.733214            7.964257    666.434583   
min         0.000000            0.000000      0.000000   
25%       687.595000            0.000000      0.000000   
50%      1389.230000            0.000000      0.000000   
75%      2842.795000            0.000000      0.000000   
max     23480.140000          180.200000  29623.350000   

       collection_recovery_fee  last_pymnt_amnt  next_pymnt_d  \
count             19908.000000     19908.000000           0.0   
mean                 11.861625 

### Paid Off Rate and Annual Return

The purpose of the project is to identify the most profitable loans. There are two criteria that help you evaluate the loan portfolio profitability, paid off rate and annual return of the portfolio.

#### Paid off rate
The dataset has loans initiated from 2007 to 2011. All loans have been either fully paid or charged off. So we will create a 'repaid' column by encoding loan status and map Charged Off to 0 and Fully Paid to 1. 

In [6]:
loan_df.loan_status.value_counts()

Fully Paid     16965
Charged Off     2943
Name: loan_status, dtype: int64

In [7]:
mapping_dict = {'Charged Off':0, 'Fully Paid':1}
loan_df['repaid'] = loan_df.loan_status.map(mapping_dict)
loan_df.repaid.value_counts()

1    16965
0     2943
Name: repaid, dtype: int64

The average repaid will be the paid off rate for a loan portfolio. For example, we get overall paid off rate of all loans in the dataset in the next code cell. About 85% of all loans are paid off.

In [8]:
loan_df.repaid.mean()

0.8521699819168174

#### Annual return

Calculation of loan return is very complicated since the loan is paid by monthly installments. In this project, we simplify the calculation by using the total payment and funded amount. For charged off loans, total payment collected includes post charge off recoveries. So we can use following formula to calculate the total return:

$TotalReturn = \frac{Total Payment + recoveries}{Funded Amount} - 1$

The overall return doesn't reflect loan profitability since loans have different terms. It's more accurate to compare annual returns. There are only two terms in the dataset, 36 months and 60 months. The formula to calculate annual return is:

$Annualized Return = (1+Total Return)^{(1/years)} - 1$. 

For example, if overall return of a 36 month loan is 10%, then annualized return = `(1 + 0.1)**(1/3) - 1` = `0.032`.

Again, this is not the true annualized return on a loan investment. But the goal of this project is to identify loans to invest, so we just need a benchmark to evaluate loan performance.

In the following code cells, we first create total_return and annual_return for each loan, then we define a function to calculate annual return of a loan portfolio. The function calculates annual return for 36 month and 60 month loans in a portfolio separately. **Note:** You don't have to use this function in your project. You can define your own ways to evaluate profitability of a portfolio.

In [9]:
loan_df.term.value_counts()

36 months    14852
60 months     5056
Name: term, dtype: int64

In [10]:
loan_df['total_return'] = (loan_df.total_pymnt + loan_df.recoveries)/loan_df.funded_amnt-1
loan_df['loan_term_year'] = loan_df.term.apply(lambda x:3 if x=='36 months' else 5)
loan_df['annual_return'] = loan_df.apply(lambda x:(1+x.total_return)**(1/x.loan_term_year)-1, axis=1)

In [11]:
loan_df.sample(5)

loan_amnt  funded_amnt  funded_amnt_inv       term  int_rate  \
7735       20000        13750     13750.000000  60 months     17.49   
6545       14700        14700     14700.000000  36 months      5.99   
18950      15000        15000      8817.735227  36 months     14.42   
9990        4975         4975      4975.000000  36 months     10.37   
18961       5000         5000      3832.539027  36 months     13.47   

       installment grade sub_grade                         emp_title  \
7735        345.36     D        D5               State of California   
6545        447.14     A        A2         montefiore medical center   
18950       515.74     D        D2  American Cleaners & Laundry Inc.   
9990        161.40     B        B3           Jefferson Clinton Hotel   
18961       169.62     C        C4             Byn flowback Services   

      emp_length home_ownership  annual_inc verification_status   issue_d  \
7735   10+ years       MORTGAGE     70800.0     Source Verified  May-2011   
6545   10+ years           RENT     37000.0            Verified  Jun-2011   
18950    8 years           RENT     70000.0            Verified  Jan-2009   
9990     5 years           RENT     30000.0     Source Verified  Feb-2011   
18961    3 years           RENT     65000.0     Source Verified  Jan-2009   

      loan_status pymnt_plan             purpose                      title  \
7735   Fully Paid          n    home_improvement           Home Improvement   
6545   Fully Paid          n  debt_consolidation                      Broke   
18950  Fully Paid          n         credit_card  Credit Card Rates To High   
9990   Fully Paid          n  debt_consolidation    Debt Consolidation Loan   
18961  Fully Paid          n  debt_consolidation                  garzaloan   

      zip_code addr_state    dti  delinq_2yrs earliest_cr_line  \
7735     925xx         CA  13.34          0.0         Jun-1994   
6545     104xx         NY  10.77          0.0         Aug-1989   
18950    934xx         CA  13.70          0.0         Jan-2002   
9990     130xx         NY  23.16          1.0         Aug-1979   
18961    790xx         TX  12.48          0.0         Dec-2000   

       inq_last_6mths  mths_since_last_delinq  mths_since_last_record  \
7735              0.0                     NaN                     NaN   
6545              0.0                     NaN                     NaN   
18950             0.0                     NaN                     NaN   
9990              0.0                    20.0                     NaN   
18961             1.0                     NaN                     NaN   

       open_acc  pub_rec  revol_bal  revol_util  total_acc  \
7735       11.0      0.0      16281        76.4       35.0   
6545       12.0      0.0      14994        33.7       25.0   
18950       5.0      0.0      11129        89.0       11.0   
9990       13.0      0.0      16348        63.5       31.0   
18961      16.0      0.0       8681        42.3       21.0   

      initial_list_status  out_prncp  out_prncp_inv   total_pymnt  \
7735                    f        0.0            0.0  18945.845959   
6545                    f        0.0            0.0  16096.877871   
18950                   f        0.0            0.0  18566.549222   
9990                    f        0.0            0.0   5585.253468   
18961                   f        0.0            0.0   6163.721321   

       total_pymnt_inv  total_rec_prncp  total_rec_int  total_rec_late_fee  \
7735          18945.85         13750.00        5195.85                 0.0   
6545          16096.88         14700.00        1396.88                 0.0   
18950         10357.79         14999.98        3566.57                 0.0   
9990           5585.25          4975.00         610.25                 0.0   
18961          4646.86          5000.00        1163.72                 0.0   

       recoveries  collection_recovery_fee last_pymnt_d  last_pymnt_amnt  \
7735          0.0                      0

In [12]:
loan_df.loan_term_year.value_counts()

3    14852
5     5056
Name: loan_term_year, dtype: int64

In [13]:
def get_portfolio_annual_return(df):
    '''
    Get annual return of 36 and 60 month loans in the portfolio.
    '''
    annual_return_36, annual_return_60 = 0, 0
    df_36 = df[df.loan_term_year==3]
    if(len(df_36)>0):
        return_36 = (df_36.total_pymnt.sum() + df_36.recoveries.sum())/df_36.funded_amnt.sum()-1
        annual_return_36 = (1+return_36)**(1/3)-1
    df_60 = df[df.loan_term_year==5]
    if(len(df_60)>0):
        return_60 = (df_60.total_pymnt.sum() + df_60.recoveries.sum())/df_60.funded_amnt.sum()-1
        annual_return_60 = (1+return_60)**(1/5)-1
    print (f'36 months loan:{len(df_36)}, Annual return:{round(annual_return_36, 4)}')
    print (f'60 months loan:{len(df_60)}, Annual return:{round(annual_return_60, 4)}')
    return annual_return_36, annual_return_60

We get annual returns of all loans in the dataset with the function in the next code cell. The 36 months loans have about 3.03% annual return and the 60 months loans have about 3.13% annual return.

In [14]:
# Overall return of all loans in the dataset
get_portfolio_annual_return(loan_df)

36 months loan:14852, Annual return:0.0303
60 months loan:5056, Annual return:0.0313


(0.03028789309763802, 0.031295585034884166)

### Confusion Matrix Function

The next code cell defines function that plot the confusion matrix. You may use the function in your model evaluation.

In [15]:
# This method produces a colored heatmap that displays the relationship
# between predicted and actual types from a machine learning method.
def confusion(test, predict, labels, title='Confusion Matrix'):
    '''
        test: true label of test data, must be one dimensional
        predict: predicted label of test data, must be one dimensional
        labels: list of label names, ie: ['positive', 'negative']
        title: plot title
    '''

    bins = len(labels)
    # Make a 2D histogram from the test and result arrays
    pts, xe, ye = np.histogram2d(test, predict, bins)

    # For simplicity we create a new DataFrame
    pd_pts = pd.DataFrame(pts.astype(int), index=labels, columns=labels )
    
    # Display heatmap and add decorations
    hm = sns.heatmap(pd_pts, annot=True, fmt="d")    
    hm.axes.set_title(title, fontsize=20)
    hm.axes.set_xlabel('Predicted', fontsize=18)
    hm.axes.set_ylabel('Actual', fontsize=18)

    return None

----

# Notebook Report Tasks (160 points)

There are four tasks you need to finish for this notebook report. Each task accounts for certain points. There are total 160 points for this notebook report.

- Data understanding and feature selection (60 points).
- Modeling (40 points).
- Construct loan portfolio (40 points).
- Conclusion (20 points).

Please include all needed python code in this notebook. Each task should have a summary. 

**Please run all code cells before this cell before proceeding.**

## Task 1: Data understanding and feature selection (60 points)

Use domain knowledge and exploratory data analysis to make the initial feature selection. You need to justify why you discard or select a feature. For example, you may say, following features are excluded since they are only available after a loan is initiated: feature1, feature2, ....

Please include all python code used to help you make the selection. You can have as may code cells as needed and you will need to summarize your selections and reasons in a markdown cell.

After task 1, you should have a new DataFrame that contains the selected features. This DataFrame should also include the three features that help to calculate loan returns, total_pymnt, recoveries and funded_amnt. You will not use the three features to select loans but you will need them to evaluate your model and portfolio.

You may use(but not limited to) following criteria to select features:

- Exclude features that are only available after loan initiation. For example, last_pymnt_amnt is not a feature you can use to help you select loans.
- Exclude features that have too many missing values **and** the missing values are hard to fill.
- Exclude categorical features that have too many categories.
- Exclude categorical features that have only one category.
- Exclude features that have just one value.
- Include features that have impact on paid off rate and return.

You may use(but not limited to) following techniques:

- Domain knowledge(common sense). For example, annual income is likely to have positive impact on paid off rate.
- Exploratory Data Analysis which includes:
    - Descriptive statistics
    - Groupby
    - Pivot table
    - Visualization

You are **not** limited or required to use the techniques listed above..

In [16]:
loan_df.columns

Index(['loan_amnt', 'funded_amnt', 'funded_amnt_inv', 'term', 'int_rate',
       'installment', 'grade', 'sub_grade', 'emp_title', 'emp_length',
       'home_ownership', 'annual_inc', 'verification_status', 'issue_d',
       'loan_status', 'pymnt_plan', 'purpose', 'title', 'zip_code',
       'addr_state', 'dti', 'delinq_2yrs', 'earliest_cr_line',
       'inq_last_6mths', 'mths_since_last_delinq', 'mths_since_last_record',
       'open_acc', 'pub_rec', 'revol_bal', 'revol_util', 'total_acc',
       'initial_list_status', 'out_prncp', 'out_prncp_inv', 'total_pymnt',
       'total_pymnt_inv', 'total_rec_prncp', 'total_rec_int',
       'total_rec_late_fee', 'recoveries', 'collection_recovery_fee',
       'last_pymnt_d', 'last_pymnt_amnt', 'next_pymnt_d', 'last_credit_pull_d',
       'collections_12_mths_ex_med', 'mths_since_last_major_derog',
       'policy_code', 'application_type', 'acc_now_delinq',
       'chargeoff_within_12_mths', 'delinq_amnt', 'pub_rec_bankruptcies',
       'tax_lie

In [17]:
def evaluate_continuous_feature(feature):
    print('feature:', feature)
    print('avg by repaid:', loan_df.groupby('repaid')[feature].mean())
    print('corr to return:', round(loan_df[[feature, 'annual_return']].corr().iat[0,1], 3))

In [18]:
def evaluate_categorical_feature(feature):
    print('feature:', feature)
    print('value counts:\n', loan_df[feature].value_counts())
    print('avg repaid:\n', loan_df.groupby(feature).repaid.mean().sort_values())
    print('avg annual return:\n', loan_df.groupby(feature).annual_return.mean().sort_values())

#### funded_amnt
lower funded_amnt higher repaid.  
No correlation to annual_return. keep.

In [19]:
evaluate_continuous_feature('funded_amnt')

feature: funded_amnt
avg by repaid: repaid
0    11904.289840
1    10920.300619
Name: funded_amnt, dtype: float64
corr to return: 0.01


#### emp_title
Too many values

In [20]:
len(loan_df.emp_title.unique())

15048

#### emp_length
Longer emp_length seems to have lower repaid rate, but not very significant.
It doesn't have clear impact on loan return.

In [21]:
evaluate_categorical_feature('emp_length')

feature: emp_length
value counts:
 10+ years    4249
< 1 year     2306
2 years      2218
3 years      2108
4 years      1871
5 years      1649
1 year       1637
6 years      1151
7 years       867
8 years       718
9 years       635
Name: emp_length, dtype: int64
avg repaid:
 emp_length
10+ years    0.837138
5 years      0.843542
6 years      0.845352
8 years      0.848189
7 years      0.850058
1 year       0.852169
4 years      0.857830
< 1 year     0.862966
3 years      0.864326
9 years      0.869291
2 years      0.874211
Name: repaid, dtype: float64
avg annual return:
 emp_length
6 years      0.016613
7 years      0.017060
8 years      0.017716
1 year       0.017930
< 1 year     0.018607
4 years      0.018954
5 years      0.019462
10+ years    0.019721
9 years      0.020286
2 years      0.023222
3 years      0.023345
Name: annual_return, dtype: float64


In [22]:
loan_df.pivot_table(index='emp_length', values='repaid', aggfunc='mean').sort_values(by='repaid')

repaid
emp_length          
10+ years   0.837138
5 years     0.843542
6 years     0.845352
8 years     0.848189
7 years     0.850058
1 year      0.852169
4 years     0.857830
< 1 year    0.862966
3 years     0.864326
9 years     0.869291
2 years     0.874211

In [23]:
for i in loan_df.emp_length.unique():
    print ('*'*10, i)
    get_portfolio_annual_return(loan_df[loan_df.emp_length==i])

********** 10+ years
36 months loan:2849, Annual return:0.0298
60 months loan:1400, Annual return:0.0313
********** 9 years
36 months loan:465, Annual return:0.0277
60 months loan:170, Annual return:0.0301
********** 4 years
36 months loan:1436, Annual return:0.0289
60 months loan:435, Annual return:0.0328
********** < 1 year
36 months loan:1857, Annual return:0.03
60 months loan:449, Annual return:0.0308
********** 3 years
36 months loan:1612, Annual return:0.0345
60 months loan:496, Annual return:0.0329
********** 6 years
36 months loan:840, Annual return:0.0308
60 months loan:311, Annual return:0.0281
********** 5 years
36 months loan:1207, Annual return:0.0302
60 months loan:442, Annual return:0.0311
********** 1 year
36 months loan:1329, Annual return:0.0294
60 months loan:308, Annual return:0.0286
********** 2 years
36 months loan:1736, Annual return:0.033
60 months loan:482, Annual return:0.0377
********** 7 years
36 months loan:617, Annual return:0.0323
60 months loan:250, Annu

#### home_ownership

Seems borrowers have home mortgage have higher repaid. But not affect loan return much.


In [24]:
loan_df.home_ownership.value_counts()

RENT        10966
MORTGAGE     7351
OWN          1546
OTHER          44
NONE            1
Name: home_ownership, dtype: int64

In [25]:
loan_df.groupby('home_ownership').repaid.mean()

home_ownership
MORTGAGE    0.861379
NONE        1.000000
OTHER       0.818182
OWN         0.841527
RENT        0.847620
Name: repaid, dtype: float64

In [26]:
get_portfolio_annual_return(loan_df[loan_df.home_ownership=='MORTGAGE'])

36 months loan:5058, Annual return:0.0301
60 months loan:2293, Annual return:0.0322


(0.030070349981624833, 0.032192668105269995)

In [27]:
get_portfolio_annual_return(loan_df[loan_df.home_ownership!='MORTGAGE'])

36 months loan:9794, Annual return:0.0304
60 months loan:2763, Annual return:0.0304


(0.030422513198133272, 0.030384409990375394)

#### verification_status
Seems verifed has lower repaid rate, but no obvious impact on return.

In [28]:
loan_df.verification_status.value_counts()

Not Verified       8257
Verified           6464
Source Verified    5187
Name: verification_status, dtype: int64

In [29]:
loan_df.groupby('verification_status').repaid.mean()

verification_status
Not Verified       0.867991
Source Verified    0.852130
Verified           0.831993
Name: repaid, dtype: float64

In [30]:
get_portfolio_annual_return(loan_df[loan_df.verification_status=='Verified'])

36 months loan:3984, Annual return:0.0318
60 months loan:2480, Annual return:0.0316


(0.031834098087079044, 0.03159903546654852)

In [31]:
get_portfolio_annual_return(loan_df[loan_df.verification_status=='Not Verified'])

36 months loan:7199, Annual return:0.0287
60 months loan:1058, Annual return:0.031


(0.028702584109791873, 0.030952124857072105)

In [32]:
get_portfolio_annual_return(loan_df[loan_df.verification_status=='Source Verified'])

36 months loan:3669, Annual return:0.0307
60 months loan:1518, Annual return:0.0308


(0.03070123312256179, 0.030753462667622466)

#### pymnt_plan
Only value 'n'.

In [33]:
loan_df.pymnt_plan.value_counts()

n    19908
Name: pymnt_plan, dtype: int64

#### purpose
small business has low repaid and very low return. 

**May not be very effective for repaid classification, but can be used to construct loan portfolio.**
Avoid:
- small_business
- educational

In [34]:
loan_df.purpose.value_counts()

debt_consolidation    9389
credit_card           2696
other                 2060
home_improvement      1330
major_purchase        1048
small_business         905
car                    687
wedding                539
medical                350
moving                 317
vacation               199
house                  176
educational            161
renewable_energy        51
Name: purpose, dtype: int64

In [35]:
loan_df.groupby('purpose').repaid.mean().sort_values()

purpose
small_business        0.728177
educational           0.801242
house                 0.823864
moving                0.829653
other                 0.843689
debt_consolidation    0.845670
medical               0.854286
vacation              0.859296
renewable_energy      0.862745
home_improvement      0.875188
credit_card           0.885757
major_purchase        0.887405
car                   0.892285
wedding               0.894249
Name: repaid, dtype: float64

In [36]:
get_portfolio_annual_return(loan_df[loan_df.purpose=='small_business'])

36 months loan:606, Annual return:0.0073
60 months loan:299, Annual return:0.0141


(0.007348532890204762, 0.014140598402353266)

In [37]:
get_portfolio_annual_return(loan_df[loan_df.purpose=='educational'])

36 months loan:153, Annual return:0.0147
60 months loan:8, Annual return:-0.0213


(0.014712612160207694, -0.02134746335636284)

In [38]:
get_portfolio_annual_return(loan_df[loan_df.purpose=='house'])

36 months loan:125, Annual return:0.019
60 months loan:51, Annual return:0.0344


(0.018977901365689442, 0.034370596744545345)

In [39]:
get_portfolio_annual_return(loan_df[loan_df.purpose=='moving'])

36 months loan:268, Annual return:0.0237
60 months loan:49, Annual return:0.034


(0.023659315664595892, 0.03396582158201311)

In [40]:
get_portfolio_annual_return(loan_df[loan_df.purpose=='credit_card'])

36 months loan:2174, Annual return:0.0362
60 months loan:522, Annual return:0.0356


(0.03615159390955358, 0.03562460449821381)

In [41]:
get_portfolio_annual_return(loan_df[loan_df.purpose=='wedding'])

36 months loan:411, Annual return:0.0309
60 months loan:128, Annual return:0.0396


(0.03094895262191022, 0.039649467792067794)

#### title
too many unique values.

In [42]:
len(loan_df.title.unique())

10539

In [43]:
len(loan_df.zip_code.unique())

260

#### addr_state

FL and CA has low repaid rate return. Consider keep.

In [44]:
evaluate_categorical_feature('addr_state')

feature: addr_state
value counts:
 CA    7105
NY    3817
FL    2872
TX    2734
NJ    1855
IL    1525
Name: addr_state, dtype: int64
avg repaid:
 addr_state
FL    0.822772
CA    0.840816
NJ    0.847978
NY    0.868745
IL    0.870164
TX    0.882224
Name: repaid, dtype: float64
avg annual return:
 addr_state
FL    0.012247
CA    0.014999
NJ    0.021213
IL    0.023692
NY    0.024185
TX    0.026185
Name: annual_return, dtype: float64


In [45]:
loan_df.groupby('addr_state').repaid.mean().sort_values()

addr_state
FL    0.822772
CA    0.840816
NJ    0.847978
NY    0.868745
IL    0.870164
TX    0.882224
Name: repaid, dtype: float64

In [46]:
get_portfolio_annual_return(loan_df[loan_df.addr_state=='FL'])

36 months loan:2119, Annual return:0.0231
60 months loan:753, Annual return:0.0259


(0.023092500988016296, 0.02585589231603236)

In [47]:
get_portfolio_annual_return(loan_df[loan_df.addr_state=='TX'])

36 months loan:2049, Annual return:0.035
60 months loan:685, Annual return:0.0373


(0.035030662713859906, 0.0373398008557857)

In [48]:
get_portfolio_annual_return(loan_df[loan_df.addr_state=='NY'])

36 months loan:2853, Annual return:0.0347
60 months loan:964, Annual return:0.0343


(0.03469840031957938, 0.03430616254694163)

In [49]:
get_portfolio_annual_return(loan_df[loan_df.addr_state=='CA'])

36 months loan:5362, Annual return:0.0276
60 months loan:1743, Annual return:0.0278


(0.02757130253848228, 0.02777533772585672)

In [50]:
get_portfolio_annual_return(loan_df[loan_df.addr_state=='NJ'])

36 months loan:1337, Annual return:0.0317
60 months loan:518, Annual return:0.0325


(0.0316573626495007, 0.03252803815514227)

#### dti

For continous features, check:
- statistics
- correlation to annual return
- groupby repaid and compare mean

Lower dti means better repaid rate. dti has negative correlation to return.

In [51]:
loan_df.dti.describe()

count    19908.000000
mean        13.008619
std          6.663658
min          0.000000
25%          7.840000
50%         13.075000
75%         18.200000
max         29.990000
Name: dti, dtype: float64

In [52]:
loan_df.groupby('repaid').dti.mean()

repaid
0    13.628270
1    12.901125
Name: dti, dtype: float64

In [53]:
loan_df[loan_df.dti<5].repaid.mean()

0.8623292595255212

In [54]:
loan_df[loan_df.dti>18].repaid.mean()

0.8295234405269275

In [55]:
get_portfolio_annual_return(loan_df[loan_df.dti<5])

36 months loan:2218, Annual return:0.0243
60 months loan:564, Annual return:0.0286


(0.02427484008133307, 0.028604369607446056)

In [56]:
get_portfolio_annual_return(loan_df[loan_df.dti<13])

36 months loan:7640, Annual return:0.0298
60 months loan:2244, Annual return:0.0322


(0.029831946529277786, 0.03222566572356844)

In [57]:
get_portfolio_annual_return(loan_df[loan_df.dti>13])

36 months loan:7201, Annual return:0.0307
60 months loan:2811, Annual return:0.0306


(0.030719504006554432, 0.030603082614490296)

In [58]:
get_portfolio_annual_return(loan_df[loan_df.dti>18])

36 months loan:3654, Annual return:0.0304
60 months loan:1508, Annual return:0.0298


(0.03041543054725926, 0.029760281366543406)

In [59]:
loan_df[['dti', 'annual_return']].corr()

dti  annual_return
dti            1.000000      -0.006038
annual_return -0.006038       1.000000

In [60]:
loan_df[['dti', 'int_rate']].corr()

dti  int_rate
dti       1.000000  0.101981
int_rate  0.101981  1.000000

In [61]:
loan_df[['int_rate', 'annual_return']].corr()

int_rate  annual_return
int_rate       1.000000      -0.003257
annual_return -0.003257       1.000000

#### delinq_2yrs
has minor impact on repaid. But delinq_2yrs>0 has much higher return. This means the interest rate increase more than fair value.

**keep for portfolio selection, may not useful for repaid classification**

In [62]:
loan_df.delinq_2yrs.value_counts()

0.0     17790
1.0      1617
2.0       341
3.0       111
4.0        29
5.0        12
7.0         3
6.0         3
11.0        1
8.0         1
Name: delinq_2yrs, dtype: int64

In [63]:
loan_df.groupby('delinq_2yrs').repaid.mean()

delinq_2yrs
0.0     0.853513
1.0     0.840445
2.0     0.844575
3.0     0.846847
4.0     0.793103
5.0     0.916667
6.0     1.000000
7.0     0.666667
8.0     0.000000
11.0    1.000000
Name: repaid, dtype: float64

In [64]:
loan_df[loan_df.delinq_2yrs==0].repaid.mean()

0.853513209668353

In [65]:
loan_df[loan_df.delinq_2yrs>0].repaid.mean()

0.8408876298394712

In [66]:
get_portfolio_annual_return(loan_df[loan_df.delinq_2yrs==0])

36 months loan:13285, Annual return:0.0297
60 months loan:4505, Annual return:0.0303


(0.02970078561526046, 0.030250880217210785)

In [67]:
get_portfolio_annual_return(loan_df[loan_df.delinq_2yrs>0])

36 months loan:1567, Annual return:0.0355
60 months loan:551, Annual return:0.0398


(0.03553678704879837, 0.03980045000907739)

In [68]:
loan_df[['delinq_2yrs', 'int_rate']].corr()

delinq_2yrs  int_rate
delinq_2yrs     1.000000  0.157728
int_rate        0.157728  1.000000

In [69]:
loan_df[['delinq_2yrs', 'annual_return']].corr()

delinq_2yrs  annual_return
delinq_2yrs       1.000000       0.008209
annual_return     0.008209       1.000000

#### earliest_cr_line
Too many values and can't find way to use it.

In [70]:
loan_df.earliest_cr_line.unique()[:5]

array(['Nov-2001', 'Feb-1996', 'Jan-2007', 'Apr-2004', 'Sep-2004'],
      dtype=object)

In [71]:
len(loan_df.earliest_cr_line.unique())

506

#### inq_last_6mths
0 has much higher repaid. make it bi-value, 0 or 1, 1 means has inq.
Higer inq_last_6mths also means lower return.

In [72]:
loan_df.inq_last_6mths.value_counts()

0.0    10017
1.0     5393
2.0     2808
3.0     1439
4.0      137
5.0       69
6.0       26
7.0       14
8.0        5
Name: inq_last_6mths, dtype: int64

In [73]:
loan_df.groupby('inq_last_6mths').repaid.mean()

inq_last_6mths
0.0    0.876710
1.0    0.834971
2.0    0.834758
3.0    0.787352
4.0    0.824818
5.0    0.797101
6.0    0.769231
7.0    0.857143
8.0    0.600000
Name: repaid, dtype: float64

In [74]:
get_portfolio_annual_return(loan_df[loan_df.inq_last_6mths==0])

36 months loan:7585, Annual return:0.0338
60 months loan:2432, Annual return:0.0335


(0.03381404761667395, 0.03353486189814969)

In [75]:
get_portfolio_annual_return(loan_df[loan_df.inq_last_6mths==1])

36 months loan:4022, Annual return:0.0282
60 months loan:1371, Annual return:0.0294


(0.028211162273794388, 0.029414285819755248)

In [76]:
get_portfolio_annual_return(loan_df[loan_df.inq_last_6mths==2])

36 months loan:2016, Annual return:0.03
60 months loan:792, Annual return:0.0341


(0.030039206190779044, 0.03412215854084999)

In [77]:
get_portfolio_annual_return(loan_df[loan_df.inq_last_6mths==3])

36 months loan:1051, Annual return:0.0178
60 months loan:388, Annual return:0.0235


(0.017801506893379893, 0.02349002239318998)

In [78]:
get_portfolio_annual_return(loan_df[loan_df.inq_last_6mths==4])

36 months loan:96, Annual return:0.0143
60 months loan:41, Annual return:0.0044


(0.014331658913136147, 0.004389319510723411)

In [79]:
get_portfolio_annual_return(loan_df[loan_df.inq_last_6mths>2])

36 months loan:1229, Annual return:0.0151
60 months loan:461, Annual return:0.0197


(0.015148336299788623, 0.019708942640229266)

<font color='red' size = '5'> Class Example </font>

#### mths_since_last_delinq
Fail to paid debt. Can switch to has delinquency or not, 0 or 1.
Previous delinq doesn't affect repaid much, but has negative impact on interest rate. So loans with previous delinq actually have better returns.

**So, don't use this for classification on repaid, but use it to construct loan portfolio**

In [80]:
loan_df.mths_since_last_delinq.value_counts()

0.0      189
31.0     126
29.0     123
23.0     123
30.0     122
        ... 
83.0       1
97.0       1
106.0      1
103.0      1
85.0       1
Name: mths_since_last_delinq, Length: 89, dtype: int64

In [81]:
get_portfolio_annual_return(loan_df[loan_df.mths_since_last_delinq.isna()])

36 months loan:9624, Annual return:0.0288
60 months loan:3286, Annual return:0.0288


(0.028792668312181968, 0.028814082915623818)

In [82]:
get_portfolio_annual_return(loan_df[loan_df.mths_since_last_delinq>=0])

36 months loan:5228, Annual return:0.0332
60 months loan:1770, Annual return:0.0361


(0.03321397906710932, 0.03610450943030585)

In [83]:
loan_df[loan_df.mths_since_last_delinq.isna()].repaid.mean()

0.8563129357087529

In [84]:
loan_df[loan_df.mths_since_last_delinq>=0].repaid.mean()

0.8445270077164905

In [85]:
loan_df[loan_df.mths_since_last_delinq>=0].int_rate.mean()

13.293543869677208

In [86]:
loan_df[loan_df.mths_since_last_delinq.isna()].int_rate.mean()

11.43716963594077

#### mths_since_last_record
Can convert to categorical feature, 0 for all missing values, 1 for >=0.
Having record has negative impace on both repaid and return.

**So keep for classfication on repaid**

In [87]:
loan_df[loan_df.mths_since_last_record.isna()].repaid.mean()

0.856516391250869

In [88]:
loan_df[loan_df.mths_since_last_record>=0].repaid.mean()

0.7849462365591398

In [89]:
loan_df[loan_df.mths_since_last_record==0].repaid.mean()

0.7796052631578947

In [90]:
get_portfolio_annual_return(loan_df[loan_df.mths_since_last_record.isna()])

36 months loan:13887, Annual return:0.0314
60 months loan:4812, Annual return:0.0315


(0.03136517589879495, 0.03152133927129297)

In [91]:
get_portfolio_annual_return(loan_df[loan_df.mths_since_last_record>=0])

36 months loan:965, Annual return:0.0138
60 months loan:244, Annual return:0.0259


(0.01376711265145203, 0.02587799046760253)

In [92]:
loan_df[loan_df.mths_since_last_record==0].repaid.mean()

0.7796052631578947

In [93]:
loan_df[loan_df.mths_since_last_record>=0].repaid.mean()

0.7849462365591398

In [94]:
loan_df[loan_df.mths_since_last_record.isna()].repaid.mean()

0.856516391250869

#### open_acc
No impace on repaid or return

In [95]:
loan_df.open_acc.describe()

count    19908.000000
mean         9.278782
std          4.414903
min          2.000000
25%          6.000000
50%          9.000000
75%         12.000000
max         44.000000
Name: open_acc, dtype: float64

In [96]:
loan_df[loan_df.open_acc<=6].repaid.mean()

0.8461277749533977

In [97]:
loan_df[loan_df.open_acc>=12].repaid.mean()

0.854762355614467

In [98]:
get_portfolio_annual_return(loan_df[loan_df.open_acc<=6])

36 months loan:4631, Annual return:0.0303
60 months loan:1270, Annual return:0.0301


(0.03029301397908113, 0.030091352966041818)

In [99]:
get_portfolio_annual_return(loan_df[loan_df.open_acc>=12])

36 months loan:3804, Annual return:0.0304
60 months loan:1477, Annual return:0.034


(0.030423885396715145, 0.034029206841818205)

#### pub_rec
Number of derogatory public records

Has impact on repaid, but not return. keep for classification.

In [100]:
evaluate_categorical_feature('pub_rec')

feature: pub_rec
value counts:
 0.0    19003
1.0      884
2.0       17
3.0        4
Name: pub_rec, dtype: int64
avg repaid:
 pub_rec
1.0    0.785068
2.0    0.823529
0.0    0.855286
3.0    1.000000
Name: repaid, dtype: float64
avg annual return:
 pub_rec
1.0    0.004657
0.0    0.019786
2.0    0.043347
3.0    0.071315
Name: annual_return, dtype: float64


#### revol_bal

a revolving balance is the portion of credit card spending that goes unpaid at the end of a billing cycle. The amount can vary, going up or down depending on the amount borrowed and the amount repaid. ... When the balance is paid off, the customer is no longer revolving the debt.

**not much impact on repaid, but higher revol_bal means higher return**  

**but why correlation is negative? because higher revol_bal higher loan amount, higher interest rate, but not much drop in repaid**

In [101]:
evaluate_continuous_feature('revol_bal')

feature: revol_bal
avg by repaid: repaid
0    13278.195719
1    13378.878809
Name: revol_bal, dtype: float64
corr to return: 0.018


In [102]:
loan_df.revol_bal.describe()

count     19908.000000
mean      13363.994826
std       15943.303849
min           0.000000
25%        3770.500000
50%        8876.500000
75%       16952.250000
max      148829.000000
Name: revol_bal, dtype: float64

In [103]:
loan_df[loan_df.revol_bal<3800].repaid.mean()

0.860423322683706

In [104]:
loan_df[loan_df.revol_bal<3800].annual_return.mean()

0.016341575962200122

In [105]:
loan_df[loan_df.revol_bal>16900].repaid.mean()

0.8450929442334599

In [106]:
loan_df[loan_df.revol_bal>20000].annual_return.mean()

0.020697149669823802

In [107]:
get_portfolio_annual_return(loan_df[loan_df.revol_bal<1000])

36 months loan:1399, Annual return:0.0213
60 months loan:392, Annual return:0.0217


(0.02133583614947754, 0.021739790517681712)

In [108]:
get_portfolio_annual_return(loan_df[loan_df.revol_bal<2000])

36 months loan:2307, Annual return:0.0235
60 months loan:613, Annual return:0.0249


(0.02346792644311968, 0.024902080994441356)

In [109]:
get_portfolio_annual_return(loan_df[loan_df.revol_bal>20000])

36 months loan:2681, Annual return:0.0339
60 months loan:1229, Annual return:0.0352


(0.03387542728906645, 0.03524243802359983)

In [110]:
get_portfolio_annual_return(loan_df[loan_df.revol_bal>100000])

36 months loan:83, Annual return:0.0458
60 months loan:38, Annual return:0.0589


(0.04582932395793082, 0.05891051221245869)

In [111]:
loan_df[['revol_bal', 'int_rate']].corr()

revol_bal  int_rate
revol_bal   1.000000  0.084654
int_rate    0.084654  1.000000

In [112]:
loan_df[['revol_bal', 'annual_return']].corr()

revol_bal  annual_return
revol_bal       1.000000       0.018403
annual_return   0.018403       1.000000

In [113]:
loan_df[['revol_bal', 'funded_amnt']].corr()

revol_bal  funded_amnt
revol_bal     1.000000     0.302797
funded_amnt   0.302797     1.000000

#### revol_util
Your revolving utilization ratio is also known as your debt-to-limit ratio or your credit utilization ratio. It measures how much of your credit limits are in use on each of your credit card accounts and expresses that calculation as a percentage.

**lower revol_util higher repaid, higher return, keep**

In [114]:
evaluate_continuous_feature('revol_util')

feature: revol_util
avg by repaid: repaid
0    55.496101
1    48.372278
Name: revol_util, dtype: float64
corr to return: 0.009


#### total_acc
Number of credit card.

**total_acc has positive impact on repaid, but a little negative impact on annual_return, avoid extremely low total_acc(<5?)**

In [115]:
#evaluate_categorical_feature('total_acc')
loan_df.total_acc.describe()

count    19908.000000
mean        21.527627
std         11.269006
min          2.000000
25%         13.000000
50%         20.000000
75%         28.000000
max         90.000000
Name: total_acc, dtype: float64

In [116]:
evaluate_continuous_feature('total_acc')

feature: total_acc
avg by repaid: repaid
0    20.761468
1    21.660536
Name: total_acc, dtype: float64
corr to return: 0.005


In [117]:
loan_df[loan_df.total_acc<10].mean()[['repaid', 'annual_return']]

repaid           0.820671
annual_return    0.015588
dtype: float64

In [118]:
get_portfolio_annual_return(loan_df[loan_df.total_acc<5])

36 months loan:260, Annual return:0.0247
60 months loan:55, Annual return:0.0135


(0.024701897998181233, 0.013508017472367628)

In [119]:
loan_df[loan_df.total_acc>30].mean()[['repaid', 'annual_return']]

repaid           0.855740
annual_return    0.017365
dtype: float64

In [120]:
get_portfolio_annual_return(loan_df[loan_df.total_acc>30])

36 months loan:2720, Annual return:0.0287
60 months loan:1252, Annual return:0.0309


(0.02867727407814069, 0.030883586380709538)

#### initial_list_status

Only f.

In [121]:
evaluate_categorical_feature('initial_list_status')

feature: initial_list_status
value counts:
 f    19908
Name: initial_list_status, dtype: int64
avg repaid:
 initial_list_status
f    0.85217
Name: repaid, dtype: float64
avg annual return:
 initial_list_status
f    0.019144
Name: annual_return, dtype: float64


#### collections_12_mths_ex_med
all 0

In [122]:
loan_df.collections_12_mths_ex_med.describe()

count    19893.0
mean         0.0
std          0.0
min          0.0
25%          0.0
50%          0.0
75%          0.0
max          0.0
Name: collections_12_mths_ex_med, dtype: float64

#### pub_rec_bankruptcies
Has negative impact on repaid and return. keep.

In [123]:
evaluate_categorical_feature('pub_rec_bankruptcies')

feature: pub_rec_bankruptcies
value counts:
 0.0    18852
1.0      742
2.0        1
Name: pub_rec_bankruptcies, dtype: int64
avg repaid:
 pub_rec_bankruptcies
1.0    0.784367
0.0    0.856090
2.0    1.000000
Name: repaid, dtype: float64
avg annual return:
 pub_rec_bankruptcies
1.0    0.000246
0.0    0.020173
2.0    0.091377
Name: annual_return, dtype: float64


In [124]:
get_portfolio_annual_return(loan_df[loan_df.pub_rec_bankruptcies==0])

36 months loan:14003, Annual return:0.0314
60 months loan:4849, Annual return:0.0316


(0.03140944706766691, 0.031617186069990755)

In [125]:
get_portfolio_annual_return(loan_df[loan_df.pub_rec_bankruptcies>0])

36 months loan:536, Annual return:0.0178
60 months loan:207, Annual return:0.0219


(0.017783268775978023, 0.0219108422355081)

#### hardship_flat, disbursement_method, debt_settlement_flag

In [126]:
loan_df.hardship_flag.unique()

array(['N'], dtype=object)

In [127]:
loan_df.disbursement_method.unique()

array(['Cash'], dtype=object)

In [128]:
loan_df.debt_settlement_flag.unique()

array(['N', 'Y'], dtype=object)

In [129]:
evaluate_categorical_feature('debt_settlement_flag')

feature: debt_settlement_flag
value counts:
 N    19826
Y       82
Name: debt_settlement_flag, dtype: int64
avg repaid:
 debt_settlement_flag
Y    0.000000
N    0.855695
Name: repaid, dtype: float64
avg annual return:
 debt_settlement_flag
N    0.019129
Y    0.022810
Name: annual_return, dtype: float64


### Task 1 Summary
Write your summary in the next markdown cell.

In [130]:
loan_df.describe()

loan_amnt   funded_amnt  funded_amnt_inv      int_rate  \
count  19908.000000  19908.000000     19908.000000  19908.000000   
mean   11353.846444  11065.763763     10500.929748     12.089717   
std     7463.700492   7176.276661      7106.229650      3.698287   
min      500.000000    500.000000         0.000000      5.420000   
25%     5750.000000   5600.000000      5000.000000      9.450000   
50%    10000.000000  10000.000000      9000.000000     11.860000   
75%    15000.000000  15000.000000     14500.000000     14.610000   
max    35000.000000  35000.000000     35000.000000     24.400000   

        installment    annual_inc           dti   delinq_2yrs  inq_last_6mths  \
count  19908.000000  1.990800e+04  19908.000000  19908.000000    19908.000000   
mean     330.614254  7.107345e+04     13.008619      0.143962        0.829466   
std      210.557434  6.980565e+04      6.663658      0.489576        1.044978   
min       15.690000  4.000000e+03      0.000000      0.000000        0.000000   
25%      171.287500  4.200000e+04      7.840000      0.000000        0.000000   
50%      285.780000  6.000000e+04     13.075000      0.000000        0.000000   
75%      440.815000  8.500000e+04     18.200000      0.000000        1.000000   
max     1302.690000  6.000000e+06     29.990000     11.000000        8.000000   

       mths_since_last_delinq  mths_since_last_record      open_acc  \
count             6998.000000             1209.000000  19908.000000   
mean                35.841097               69.354839      9.278782   
std                 21.599170               44.520279      4.414903   
min                  0.000000                0.000000      2.000000   
25%                 19.000000                0.000000      6.000000   
50%                 34.000000               90.000000      9.000000   
75%                 51.000000              104.000000     12.000000   
max                106.000000              129.000000     44.000000   

            pub_rec      revol_bal    revol_util     total_acc  out_prncp  \
count  19908.000000   19908.000000  19889.000000  19908.000000    19908.0   
mean       0.046715   13363.994826     49.424966     21.527627        0.0   
std        0.217827   15943.303849     28.198395     11.269006        0.0   
min        0.000000       0.000000      0.000000      2.000000        0.0   
25%        0.000000    3770.500000     26.500000     13.000000        0.0   
50%        0.000000    8876.500000     50.300000     20.000000        0.0   
75%        0.000000   16952.250000     72.800000     28.000000        0.0   
max        3.000000  148829.000000     99.900000     90.000000        0.0   

       out_prncp_inv   total_pymnt  total_pymnt_inv  total_rec_prncp  \
count        19908.0  19908.000000     19908.000000     19908.000000   
mean             0.0  12286.852391     11690.155107      9913.519990   
std              0.0   9098.847567      8984.722396      7126.424699   
min              0.0      0.000000         0.000000         0.000000   
25%              0.0   5670.592545      5239.857500      4800.000000   
50%              0.0  10042.735817      9427.845000      8000.000000   
75%              0.0  16681.688450     15983.440000     14000.000000   
max              0.0  58480.139915     58438.370000     35000.020000   

       total_rec_int  total_rec_late_fee    recoveries  \
count   19908.000000        19908.000000  19908.000000   
mean     2277.120129            1.489973     94.722382   
std      2583.733214            7.964257    666.434583   
min         0.000000            0.000000      0.000000   
25%       687.595000            0.000000      0.000000   
50%      1389.230000            0.000000      0.000000   
75%      2842.795000            0.000000      0.000000   
max     23480.140000          180.200000  29623.350000   

       collection_recovery_fee  last_pymnt_amnt  next_pymnt_d  \
count             19908.000000     19908.000000           0.0   
mean                 11.861625 

In [131]:
loan_df.columns

Index(['loan_amnt', 'funded_amnt', 'funded_amnt_inv', 'term', 'int_rate',
       'installment', 'grade', 'sub_grade', 'emp_title', 'emp_length',
       'home_ownership', 'annual_inc', 'verification_status', 'issue_d',
       'loan_status', 'pymnt_plan', 'purpose', 'title', 'zip_code',
       'addr_state', 'dti', 'delinq_2yrs', 'earliest_cr_line',
       'inq_last_6mths', 'mths_since_last_delinq', 'mths_since_last_record',
       'open_acc', 'pub_rec', 'revol_bal', 'revol_util', 'total_acc',
       'initial_list_status', 'out_prncp', 'out_prncp_inv', 'total_pymnt',
       'total_pymnt_inv', 'total_rec_prncp', 'total_rec_int',
       'total_rec_late_fee', 'recoveries', 'collection_recovery_fee',
       'last_pymnt_d', 'last_pymnt_amnt', 'next_pymnt_d', 'last_credit_pull_d',
       'collections_12_mths_ex_med', 'mths_since_last_major_derog',
       'policy_code', 'application_type', 'acc_now_delinq',
       'chargeoff_within_12_mths', 'delinq_amnt', 'pub_rec_bankruptcies',
       'tax_lie

#### Label
loan_status
#### Columns after loan initiation
issue_d, debt_settlement_flag, debt_settlement_flag_date
#### Constant value
pymnt_plan, application_type, initial_list_status, collections_12_mths_ex_med, policy_code, acc_now_delinq, chargeoff_within_12_mths, delinq_amnt, tax_liens, hardship_flat, disbursement_method
#### Categorical feature with too many values
title, zip_code, earliest_cr_line
#### Too many missing values?
mths_since_last_major_derog
#### No impact on repaid or return
emp_length, open_acc
#### Keep all purpose
funded_amnt, loan_term_year, int_rate, grade, home_ownership, annual_inc, verification_status, purpose, addr_state, dti, delinq_2yrs, inq_last_6mths, mths_since_last_delinq, mths_since_last_record, pub_rec, revol_bal, revol_util
loan_status, repaid, annual_return, total_pymnt, recoveries, pub_rec_bankruptcies
#### Keep for classification
funded_amnt, loan_term_year, int_rate, grade, emp_length, home_ownership, annual_inc, verification_status, purpose, addr_state, dti, delinq_2yrs, inq_last_6mths, mths_since_last_record, pub_rec, revol_util, pub_rec_bankruptcies


In [132]:
keep_columns = ['funded_amnt', 'loan_term_year', 'int_rate', 'grade', 'emp_length', 'home_ownership', 'annual_inc', 
                'verification_status', 'purpose', 'addr_state', 'dti', 'delinq_2yrs', 'inq_last_6mths', 
                'mths_since_last_delinq', 'mths_since_last_record', 'pub_rec', 'revol_bal', 'revol_util', 
                'loan_status', 'repaid', 'annual_return', 'total_pymnt', 'recoveries', 'pub_rec_bankruptcies']
classification_columns = ['funded_amnt', 'loan_term_year', 'int_rate', 'grade', 'emp_length', 'home_ownership', 
                          'annual_inc', 'verification_status', 'purpose', 'addr_state', 'dti', 'delinq_2yrs', 
                          'inq_last_6mths', 'mths_since_last_record', 'pub_rec', 'revol_util', 'pub_rec_bankruptcies']
#used to calculate returns
calculate_columns = [c for c in keep_columns if c not in classification_columns]
cat_columns = ['loan_term_year','grade','emp_length','home_ownership','verification_status','purpose','addr_state']
cont_columns = ['funded_amnt','int_rate','annual_inc', 'dti', 'revol_util']
#convert to bi-value, 0 and 1
convert_to_bi_cat_columns = ['delinq_2yrs','inq_last_6mths','mths_since_last_record', 'pub_rec', 'pub_rec_bankruptcies']

In [133]:
[c for c in keep_columns if c not in classification_columns]

['mths_since_last_delinq',
 'revol_bal',
 'loan_status',
 'repaid',
 'annual_return',
 'total_pymnt',
 'recoveries']

In [134]:
#df with all useful columns
df = loan_df[keep_columns]
df['delinq_2yrs_bi'] = 0
df.loc[df.delinq_2yrs>0, 'delinq_2yrs_bi'] = 1
df['inq_last_6mths_bi'] = 0
df.loc[df.inq_last_6mths>0, 'inq_last_6mths_bi'] = 1
df['mths_since_last_record_bi'] = 0
df.loc[df.mths_since_last_record>=0, 'mths_since_last_record_bi'] = 1
df['pub_rec_bi'] = 0
df.loc[df.pub_rec>0, 'pub_rec_bi'] = 1
df['pub_rec_bankruptcies_bi'] = 0
df.loc[df.pub_rec_bankruptcies>0, 'pub_rec_bankruptcies_bi'] = 1
converted_bi_columns = ['delinq_2yrs_bi','inq_last_6mths_bi','mths_since_last_record_bi', 'pub_rec_bi', 'pub_rec_bankruptcies_bi']


## Task 2: Modeling (40 points)

Construct a classification model to predict whether a loan will be paid in full. Column 'repaid' will be the label. The purpose of the model is to identify loans that are more likely to be paid in full.

#### Preprocessing

- Separate label and data. Label is 'repaid' column. 
- Encode categorical features
- Manage missing values(drop or fill), please justify your choice.
- Split dataset to train and test. Please set random_state to ensure reproducibility.

#### Modeling, Evaluation and Optimization

- Apply at least 2 classifiers including LogisticRegression.(The reason is explained in Task3).
- If you choose Support Vector Machine, use LinearSVC instead of SVC. The reason is that SVC is very slow with this dataset.
- Apply feature selection with following techniques (You don't have to use all of them).
    - Filter methods
    - Wrapper methods
    - Embedded methods
- Optimize each classifier with cross validation. Select best options for major hyperparameters of the models you choose.
- Evaluate models with classification report and confusion matrix.
- Compare models with ROC plot.

#### Conclusion
Summarize your analysis in task 2.


### Preprocessing
#### Encode Categorical Features

In [135]:
mapping_dict = {3:0, 5:1}
#Encode term
df['term_code'] = df.loan_term_year.map(mapping_dict)
mapping_dict = {'A':0, 'B':1,'C':2,'D':3,'E':4,'F':5,'G':6}
df['grade_code'] = df.grade.map(mapping_dict)
mapping_dict = {
 '< 1 year':0,
 '1 year':1,
 '2 years':2,
 '3 years':3,
 '4 years':4,
 '5 years':5,
 '6 years':6,
 '7 years':7,
 '8 years':8,
 '9 years':9,
 '10+ years':10}
df['emp_length_code'] = df.emp_length.map(mapping_dict)
mapping_dict = {'Not Verified':0, 'Source Verified':1, 'Verified':2}
df['verification_status_code'] = df.verification_status.map(mapping_dict)

In [136]:
from sklearn.preprocessing import LabelEncoder

le = LabelEncoder()
df['purpose_code'] = le.fit_transform(df.purpose)
df['home_ownership_code'] = le.fit_transform(df.home_ownership)
df['addr_state_code'] = le.fit_transform(df.addr_state)


In [137]:
cat_columns_encoded = ['term_code','grade_code','emp_length_code','home_ownership_code',
                       'verification_status_code','purpose_code','addr_state_code']


In [138]:
model_columns = cat_columns_encoded + converted_bi_columns + cont_columns
model_columns

['term_code',
 'grade_code',
 'emp_length_code',
 'home_ownership_code',
 'verification_status_code',
 'purpose_code',
 'addr_state_code',
 'delinq_2yrs_bi',
 'inq_last_6mths_bi',
 'mths_since_last_record_bi',
 'pub_rec_bi',
 'pub_rec_bankruptcies_bi',
 'funded_amnt',
 'int_rate',
 'annual_inc',
 'dti',
 'revol_util']

In [139]:
df[model_columns].info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 19908 entries, 0 to 19907
Data columns (total 17 columns):
term_code                    19908 non-null int64
grade_code                   19908 non-null int64
emp_length_code              19409 non-null float64
home_ownership_code          19908 non-null int64
verification_status_code     19908 non-null int64
purpose_code                 19908 non-null int64
addr_state_code              19908 non-null int64
delinq_2yrs_bi               19908 non-null int64
inq_last_6mths_bi            19908 non-null int64
mths_since_last_record_bi    19908 non-null int64
pub_rec_bi                   19908 non-null int64
pub_rec_bankruptcies_bi      19908 non-null int64
funded_amnt                  19908 non-null int64
int_rate                     19908 non-null float64
annual_inc                   19908 non-null float64
dti                          19908 non-null float64
revol_util                   19889 non-null float64
dtypes: float64(5), int64(12)
m

#### Missing Values
Fill missing values

In [140]:
df.emp_length_code.value_counts()

10.0    4249
0.0     2306
2.0     2218
3.0     2108
4.0     1871
5.0     1649
1.0     1637
6.0     1151
7.0      867
8.0      718
9.0      635
Name: emp_length_code, dtype: int64

In [141]:
df.revol_util.describe()

count    19889.000000
mean        49.424966
std         28.198395
min          0.000000
25%         26.500000
50%         50.300000
75%         72.800000
max         99.900000
Name: revol_util, dtype: float64

In [142]:
df.emp_length_code.fillna(10.0, inplace=True)
df.revol_util.fillna(df.revol_util.mean(), inplace=True)
df[model_columns].info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 19908 entries, 0 to 19907
Data columns (total 17 columns):
term_code                    19908 non-null int64
grade_code                   19908 non-null int64
emp_length_code              19908 non-null float64
home_ownership_code          19908 non-null int64
verification_status_code     19908 non-null int64
purpose_code                 19908 non-null int64
addr_state_code              19908 non-null int64
delinq_2yrs_bi               19908 non-null int64
inq_last_6mths_bi            19908 non-null int64
mths_since_last_record_bi    19908 non-null int64
pub_rec_bi                   19908 non-null int64
pub_rec_bankruptcies_bi      19908 non-null int64
funded_amnt                  19908 non-null int64
int_rate                     19908 non-null float64
annual_inc                   19908 non-null float64
dti                          19908 non-null float64
revol_util                   19908 non-null float64
dtypes: float64(5), int64(12)
m

#### Split dataset

In [143]:
from sklearn.model_selection import train_test_split
df_train, df_test = train_test_split(df, test_size=0.4, random_state=23)
d_train = df_train[model_columns]
d_test = df_test[model_columns]
l_train = df_train.repaid
l_test = df_test.repaid

### Feature Selection
#### Filter Method

###### Statistical

In [144]:
from sklearn.preprocessing import MinMaxScaler
from sklearn.feature_selection import VarianceThreshold
vt = VarianceThreshold()

d_train_ss = MinMaxScaler().fit_transform(d_train)

# Compute and display variances
vt.fit_transform(d_train_ss)

for name, var in sorted(zip(d_train.columns, vt.variances_), key=lambda x: x[1], reverse=True):
    print(f'{name:>18} variance = {var:5.3f}')

 inq_last_6mths_bi variance = 0.250
home_ownership_code variance = 0.222
         term_code variance = 0.190
verification_status_code variance = 0.184
   addr_state_code variance = 0.147
   emp_length_code variance = 0.128
    delinq_2yrs_bi variance = 0.093
        revol_util variance = 0.080
      purpose_code variance = 0.071
mths_since_last_record_bi variance = 0.058
        grade_code variance = 0.054
               dti variance = 0.049
       funded_amnt variance = 0.044
        pub_rec_bi variance = 0.044
          int_rate variance = 0.038
pub_rec_bankruptcies_bi variance = 0.036
        annual_inc variance = 0.001


In [145]:
d_train.inq_last_6mths_bi.value_counts()

0    5997
1    5947
Name: inq_last_6mths_bi, dtype: int64

###### Univariate Technics

In [146]:
d_train.mths_since_last_record_bi.value_counts()

0    11207
1      737
Name: mths_since_last_record_bi, dtype: int64

In [147]:
from sklearn.feature_selection import SelectKBest

skb = SelectKBest(k='all')
fs = skb.fit(d_train, l_train)
s = '['
for var, name in sorted(zip(fs.scores_, d_train.columns), key=lambda x: x[0], reverse=True):
    print(f'{name:>18} score = {var:5.3f}')
    s += f'"{name}",'
print (s+']')

          int_rate score = 494.632
        grade_code score = 462.357
         term_code score = 294.032
        revol_util score = 94.424
 inq_last_6mths_bi score = 76.877
        annual_inc score = 24.345
   addr_state_code score = 18.607
      purpose_code score = 18.414
mths_since_last_record_bi score = 18.085
       funded_amnt score = 17.934
verification_status_code score = 17.748
        pub_rec_bi score = 12.417
   emp_length_code score = 11.772
pub_rec_bankruptcies_bi score = 11.492
               dti score = 10.211
home_ownership_code score = 2.724
    delinq_2yrs_bi score = 0.021
["int_rate","grade_code","term_code","revol_util","inq_last_6mths_bi","annual_inc","addr_state_code","purpose_code","mths_since_last_record_bi","funded_amnt","verification_status_code","pub_rec_bi","emp_length_code","pub_rec_bankruptcies_bi","dti","home_ownership_code","delinq_2yrs_bi",]


In [148]:
kbest_all = ["int_rate","grade_code","term_code","revol_util","inq_last_6mths_bi","annual_inc",
             "addr_state_code","purpose_code","mths_since_last_record_bi","funded_amnt",
             "verification_status_code","pub_rec_bi","emp_length_code","pub_rec_bankruptcies_bi",
             "dti","home_ownership_code","delinq_2yrs_bi"]
kbest_15 = ["int_rate","grade_code","term_code","revol_util","inq_last_6mths_bi","annual_inc",
             "addr_state_code","purpose_code","mths_since_last_record_bi","funded_amnt",
             "verification_status_code","pub_rec_bi","emp_length_code","pub_rec_bankruptcies_bi",
             "dti"]

#### Wrapper methods

In [149]:
from sklearn.feature_selection import RFE
from sklearn.linear_model import LogisticRegression

lr = LogisticRegression(C=1E6, class_weight='balanced')

# Perform RFE, select top feature (but rank all)
rfe = RFE(lr, 1)

# Fit features plus noise
fs = rfe.fit(d_train, l_train)
    
# Display feature ranking
for var, name in sorted(zip(fs.ranking_, d_train.columns), key=lambda x: x[0]):
    print(f'{name:>18} rank = {var}')

         term_code rank = 1
mths_since_last_record_bi rank = 2
 inq_last_6mths_bi rank = 3
        pub_rec_bi rank = 4
    delinq_2yrs_bi rank = 5
          int_rate rank = 6
pub_rec_bankruptcies_bi rank = 7
   addr_state_code rank = 8
        grade_code rank = 9
verification_status_code rank = 10
      purpose_code rank = 11
home_ownership_code rank = 12
   emp_length_code rank = 13
               dti rank = 14
        revol_util rank = 15
       funded_amnt rank = 16
        annual_inc rank = 17


In [150]:
lr = LogisticRegression()
# Perform RFE, select top feature (but rank all)
rfe = RFE(lr, 1)

# Fit features plus noise
fs = rfe.fit(d_train, l_train)
    
# Display feature ranking
for var, name in sorted(zip(fs.ranking_, d_train.columns), key=lambda x: x[0]):
    print(f'{name:>18} rank = {var}')

         term_code rank = 1
mths_since_last_record_bi rank = 2
 inq_last_6mths_bi rank = 3
        pub_rec_bi rank = 4
pub_rec_bankruptcies_bi rank = 5
    delinq_2yrs_bi rank = 6
          int_rate rank = 7
   addr_state_code rank = 8
verification_status_code rank = 9
      purpose_code rank = 10
home_ownership_code rank = 11
   emp_length_code rank = 12
        grade_code rank = 13
               dti rank = 14
        revol_util rank = 15
        annual_inc rank = 16
       funded_amnt rank = 17


In [151]:
top_10 = ['term_code',
'mths_since_last_record_bi',
 'inq_last_6mths_bi',
        'pub_rec_bi',
'pub_rec_bankruptcies_bi',
    'delinq_2yrs_bi',
          'int_rate',
   'addr_state_code',
'verification_status_code',
      'purpose_code']
top_15 = top_10 + ['home_ownership_code','emp_length_code','grade_code','dti','revol_util']

##### Compare LogisticRegression models
Try LR on all features and selected features.

In [152]:
def get_pred_return(df_test, pred):
    '''
    get returns of predicted good loans
    '''
    df_pred = df_test.copy()
    df_pred['pred'] = pred
    print("0:36month, 1:60month\n", df_pred[df_pred.pred==1].loan_term_year.value_counts())
    print("Repaid Rate:")
    print(df_pred[df_pred.pred==1].groupby('loan_term_year').repaid.mean())
    print('predicted:', get_portfolio_annual_return(df_pred[df_pred.pred==1]))
    print('all:', get_portfolio_annual_return(df_test))

In [153]:
#Train on all features
from sklearn.linear_model import LogisticRegression
from sklearn import metrics

lr = LogisticRegression(C=1E6, class_weight='balanced')
#lr = LogisticRegression(C=1E6)

lr = lr.fit(d_train, l_train)
print(lr.score(d_test, l_test))
pred = lr.predict(d_test)
print(metrics.classification_report(l_test, pred))
get_pred_return(df_test, pred)


0.6406328478151683
              precision    recall  f1-score   support

           0       0.23      0.61      0.33      1186
           1       0.90      0.65      0.75      6778

    accuracy                           0.64      7964
   macro avg       0.57      0.63      0.54      7964
weighted avg       0.80      0.64      0.69      7964

0:36month, 1:60month
 3    4483
5     369
Name: loan_term_year, dtype: int64
Repaid Rate:
loan_term_year
3    0.907874
5    0.850949
Name: repaid, dtype: float64
36 months loan:4483, Annual return:0.0297
60 months loan:369, Annual return:0.033
predicted: (0.029680725547080122, 0.03295071372595482)
36 months loan:5960, Annual return:0.0285
60 months loan:2004, Annual return:0.0303
all: (0.028477063567154515, 0.030293000871747466)


In [154]:
# Train on kbest_15
lr = LogisticRegression(C=1E6, class_weight='balanced')
lr = lr.fit(d_train[kbest_15], l_train)
print(lr.score(d_test[kbest_15], l_test))
pred = lr.predict(d_test[kbest_15])
print(metrics.classification_report(l_test, pred))
get_pred_return(df_test, pred)


0.6352335509794074
              precision    recall  f1-score   support

           0       0.23      0.61      0.33      1186
           1       0.90      0.64      0.75      6778

    accuracy                           0.64      7964
   macro avg       0.57      0.63      0.54      7964
weighted avg       0.80      0.64      0.69      7964

0:36month, 1:60month
 3    4461
5     328
Name: loan_term_year, dtype: int64
Repaid Rate:
loan_term_year
3    0.907644
5    0.859756
Name: repaid, dtype: float64
36 months loan:4461, Annual return:0.0297
60 months loan:328, Annual return:0.0332
predicted: (0.02966037716988712, 0.03319164407102071)
36 months loan:5960, Annual return:0.0285
60 months loan:2004, Annual return:0.0303
all: (0.028477063567154515, 0.030293000871747466)


In [155]:
# Train on top 10 features
lr = lr.fit(d_train[top_10], l_train)
print(lr.score(d_test[top_10], l_test))
pred = lr.predict(d_test[top_10])
print(metrics.classification_report(l_test, pred))
get_pred_return(df_test, pred)


0.6351079859367152
              precision    recall  f1-score   support

           0       0.22      0.58      0.32      1186
           1       0.90      0.64      0.75      6778

    accuracy                           0.64      7964
   macro avg       0.56      0.61      0.54      7964
weighted avg       0.80      0.64      0.69      7964

0:36month, 1:60month
 3    4482
5     380
Name: loan_term_year, dtype: int64
Repaid Rate:
loan_term_year
3    0.903614
5    0.834211
Name: repaid, dtype: float64
36 months loan:4482, Annual return:0.0294
60 months loan:380, Annual return:0.0278
predicted: (0.029406913768269094, 0.027792971206310968)
36 months loan:5960, Annual return:0.0285
60 months loan:2004, Annual return:0.0303
all: (0.028477063567154515, 0.030293000871747466)


In [156]:
#top 15
lr = lr.fit(d_train[top_15], l_train)
print(lr.score(d_test[top_15], l_test))
pred = lr.predict(d_test[top_15])
print(metrics.classification_report(l_test, pred))
get_pred_return(df_test, pred)


0.6352335509794074
              precision    recall  f1-score   support

           0       0.22      0.59      0.33      1186
           1       0.90      0.64      0.75      6778

    accuracy                           0.64      7964
   macro avg       0.56      0.62      0.54      7964
weighted avg       0.80      0.64      0.69      7964

0:36month, 1:60month
 3    4457
5     386
Name: loan_term_year, dtype: int64
Repaid Rate:
loan_term_year
3    0.905317
5    0.836788
Name: repaid, dtype: float64
36 months loan:4457, Annual return:0.0295
60 months loan:386, Annual return:0.025
predicted: (0.0295449926190412, 0.02500636031490555)
36 months loan:5960, Annual return:0.0285
60 months loan:2004, Annual return:0.0303
all: (0.028477063567154515, 0.030293000871747466)


#### Embedded method

In [157]:
from sklearn.ensemble import RandomForestClassifier
from sklearn import metrics

# Create Regressor with default properties
rf = RandomForestClassifier(random_state=23)

# Fit estimator and display score
rf = rf.fit(d_train, l_train)
print(rf.score(d_test, l_test))
pred = rf.predict(d_test)
print(metrics.classification_report(l_test, pred))

0.8286037167252637
              precision    recall  f1-score   support

           0       0.27      0.09      0.14      1186
           1       0.86      0.96      0.90      6778

    accuracy                           0.83      7964
   macro avg       0.57      0.52      0.52      7964
weighted avg       0.77      0.83      0.79      7964



In [158]:
rf.feature_importances_
for val, name in sorted(zip(rf.feature_importances_, d_train.columns), 
                        key=lambda x: x[0], reverse=True):
    print(f'{name:>18}: {100.0*val:05.2f}%')

               dti: 14.47%
        revol_util: 13.77%
        annual_inc: 13.56%
          int_rate: 12.34%
       funded_amnt: 11.68%
   emp_length_code: 07.24%
      purpose_code: 06.22%
   addr_state_code: 05.01%
        grade_code: 03.87%
verification_status_code: 03.03%
home_ownership_code: 02.59%
         term_code: 01.69%
 inq_last_6mths_bi: 01.53%
    delinq_2yrs_bi: 01.21%
mths_since_last_record_bi: 00.81%
        pub_rec_bi: 00.52%
pub_rec_bankruptcies_bi: 00.47%


#### Don't use bi columns in classification
All bi columns are highly imbalanced. df.feature_importances_ shows they are not important.

In [159]:
model_columns_nobi = cat_columns_encoded + cont_columns

from sklearn.model_selection import train_test_split
df_train, df_test = train_test_split(df, test_size=0.4, random_state=23)
d_train = df_train[model_columns_nobi]
d_test = df_test[model_columns_nobi]
l_train = df_train.repaid
l_test = df_test.repaid

In [160]:

# Create Regressor with default properties
rf = RandomForestClassifier(class_weight='balanced', random_state=23)

# Fit estimator and display score
rf = rf.fit(d_train, l_train)
print(rf.score(d_test, l_test))
pred = rf.predict(d_test)
print(metrics.classification_report(l_test, pred))

0.8333751883475641
              precision    recall  f1-score   support

           0       0.26      0.07      0.11      1186
           1       0.86      0.97      0.91      6778

    accuracy                           0.83      7964
   macro avg       0.56      0.52      0.51      7964
weighted avg       0.77      0.83      0.79      7964



In [161]:
rf.feature_importances_
for val, name in sorted(zip(rf.feature_importances_, d_train.columns), 
                        key=lambda x: x[0], reverse=True):
    print(f'{name:>18}: {100.0*val:05.2f}%')

          int_rate: 15.85%
        annual_inc: 14.41%
        revol_util: 13.85%
               dti: 13.76%
       funded_amnt: 11.15%
   emp_length_code: 06.74%
      purpose_code: 06.47%
   addr_state_code: 05.54%
        grade_code: 04.20%
verification_status_code: 03.10%
home_ownership_code: 02.54%
         term_code: 02.39%


#### Only select >5% columns

In [162]:
selected_columns = ['revol_util', 'dti', 'annual_inc', 'int_rate', 'funded_amnt', 'emp_length_code', 'purpose_code']

df_train, df_test = train_test_split(df, test_size=0.4, random_state=23)
d_train = df_train[selected_columns]
d_test = df_test[selected_columns]
l_train = df_train.repaid
l_test = df_test.repaid

# Create Regressor with default properties
rf = RandomForestClassifier(random_state=23)

# Fit estimator and display score
rf = rf.fit(d_train, l_train)
print(rf.score(d_test, l_test))
pred = rf.predict(d_test)
print(metrics.classification_report(l_test, pred))

0.8299849321948769
              precision    recall  f1-score   support

           0       0.28      0.09      0.14      1186
           1       0.86      0.96      0.91      6778

    accuracy                           0.83      7964
   macro avg       0.57      0.52      0.52      7964
weighted avg       0.77      0.83      0.79      7964



In [163]:
from sklearn.linear_model import LogisticRegression

#lr = LogisticRegression(C=1E6, class_weight='balanced')
lr = LogisticRegression(C=1E6, class_weight={1:0.1, 0:0.9})

lr = lr.fit(d_train, l_train)
print(lr.score(d_test, l_test))
pred = lr.predict(d_test)
print(metrics.classification_report(l_test, pred))
get_pred_return(df_test, pred)


0.2737317930688096
              precision    recall  f1-score   support

           0       0.16      0.95      0.28      1186
           1       0.95      0.16      0.27      6778

    accuracy                           0.27      7964
   macro avg       0.56      0.55      0.27      7964
weighted avg       0.83      0.27      0.27      7964

0:36month, 1:60month
 3    1002
5     112
Name: loan_term_year, dtype: int64
Repaid Rate:
loan_term_year
3    0.949102
5    0.919643
Name: repaid, dtype: float64
36 months loan:1002, Annual return:0.0271
60 months loan:112, Annual return:0.0423
predicted: (0.027099604233616015, 0.0423093369295926)
36 months loan:5960, Annual return:0.0285
60 months loan:2004, Annual return:0.0303
all: (0.028477063567154515, 0.030293000871747466)


#### One hot encoding

In [164]:
model_columns = cat_columns_encoded + converted_bi_columns + cont_columns
import patsy as pts 

formula = 'repaid ~ '
for c in cat_columns_encoded:
    formula += f'C({c})+'
for c in converted_bi_columns + cont_columns:
    formula += f'{c}+'
formula = formula[:-1]
formula
y, x = pts.dmatrices(formula, data=df, return_type='dataframe')


In [165]:
from sklearn.model_selection import train_test_split
df_train, df_test = train_test_split(df, test_size=0.4, random_state=23)
d_train = df_train[model_columns]
d_test = df_test[model_columns]
l_train = df_train.repaid
l_test = df_test.repaid

### Task 2 Summary

## Task 3: Construct Loan Portfolio (40 points)

Construct a loan portfolio out of the test set with the help of your classification model. The portfolio can be a small subset of all loans in the dataset, for example, 50% of all loans in the test set.

For this task, finish following mini tasks.
- Predict on the test set with optimized model.
- Select loans that are predicted to be paid off by the model.
- Calculate paid off rate of the selected loans. Is the paid off rate better than the overall paid off rate(85.2%)?
- Calculate annual return of selected loans.
- Assess the risk of your portfolio. You may use the variance of annual return as a risk benchmark.
- Adjust class_weight to 'balanced' and other options, ie. class_weight={1:0.2, 0:0.8} and use the new model to construct loan portfolio. How is the accuracy rate change? How is the paid off rate and return of the new portfolio?
- Discuss other possible ways to construct better loan portfolio. For example, what's the impact of only selecting loans from borrowers that don't have bankruptcy record(pub_rec_bankruptcies==0)?
- Summarize your finding and the performance of your portfolio.

**Note:**  
Some machine learning models are not sensitive to class_weight change. You may change some hyperparameter values to make the model more responsive to class_weight change. For example, set max_depth to 10 for a RandomForestClassifier. This is an advanced topic and you are not required to do it. Out of the models we learned, LogisticRegression is most sensitive to class_weight change. That's why you are asked to include LogisticRegression in your modeling choices in task 2.

#### Conclusion
Summarsize your analysis in task 3. 
- Is the classification model helpful in choosing portfolio? 
- Does a portfolio with higher paid off rate always have higher return?
- What other ways you can use to help you choose a portfolio with higher return and lower risk?

In [166]:
def get_pred_return(df_test, pred):
    '''
    get returns of predicted good loans
    '''
    df_pred = df_test.copy()
    df_pred['pred'] = pred
    print("0:36month, 1:60month\n", df_pred[df_pred.pred==1].loan_term_year.value_counts())
    print("Repaid Rate:")
    print(df_pred[df_pred.pred==1].groupby('loan_term_year').repaid.mean())
    print('predicted:', get_portfolio_annual_return(df_pred[df_pred.pred==1]))
    print('all:', get_portfolio_annual_return(df_test))

In [167]:
# Create Regressor with default properties
rf = RandomForestClassifier(max_depth = 5, class_weight='balanced', random_state=23)

# Fit estimator and display score
rf = rf.fit(d_train, l_train)
print(rf.score(d_test, l_test))
pred = rf.predict(d_test)
print(metrics.classification_report(l_test, pred))

0.6363636363636364
              precision    recall  f1-score   support

           0       0.23      0.59      0.33      1186
           1       0.90      0.64      0.75      6778

    accuracy                           0.64      7964
   macro avg       0.56      0.62      0.54      7964
weighted avg       0.80      0.64      0.69      7964



In [168]:
get_pred_return(df_test, pred)

0:36month, 1:60month
 3    4518
5     328
Name: loan_term_year, dtype: int64
Repaid Rate:
loan_term_year
3    0.905046
5    0.838415
Name: repaid, dtype: float64
36 months loan:4518, Annual return:0.0289
60 months loan:328, Annual return:0.0237
predicted: (0.028947407101345135, 0.023740862940193397)
36 months loan:5960, Annual return:0.0285
60 months loan:2004, Annual return:0.0303
all: (0.028477063567154515, 0.030293000871747466)


In [169]:
df_test.groupby('loan_term_year').repaid.mean()

loan_term_year
3    0.881544
5    0.760479
Name: repaid, dtype: float64

In [170]:
df_test.loan_term_year.value_counts()

3    5960
5    2004
Name: loan_term_year, dtype: int64

#### Separate 36 and 60 month loans

In [171]:
from sklearn.model_selection import train_test_split
df_36 = df[df.loan_term_year==3]
df_train, df_test = train_test_split(df_36, test_size=0.4, random_state=23)
d_train = df_train[selected_columns]
d_test = df_test[selected_columns]
l_train = df_train.repaid
l_test = df_test.repaid

In [172]:
# Create Regressor with default properties
rf = RandomForestClassifier(max_depth = 5, class_weight='balanced', random_state=23)

# Fit estimator and display score
rf = rf.fit(d_train, l_train)
print(rf.score(d_test, l_test))
pred = rf.predict(d_test)
print(metrics.classification_report(l_test, pred))

0.6224541323009595
              precision    recall  f1-score   support

           0       0.18      0.60      0.27       697
           1       0.92      0.63      0.75      5244

    accuracy                           0.62      5941
   macro avg       0.55      0.61      0.51      5941
weighted avg       0.83      0.62      0.69      5941



In [173]:
get_pred_return(df_test, pred)

0:36month, 1:60month
 3    3559
Name: loan_term_year, dtype: int64
Repaid Rate:
loan_term_year
3    0.921607
Name: repaid, dtype: float64
36 months loan:3559, Annual return:0.0295
60 months loan:0, Annual return:0
predicted: (0.029508887476334067, 0)
36 months loan:5941, Annual return:0.0302
60 months loan:0, Annual return:0
all: (0.03015380427611003, 0)


In [174]:
df_test['pred'] = pred
df_test[df_test.pred==1].annual_return.mean()

0.022542417710611662

In [175]:
df_test.repaid.mean()

0.8826796835549571

In [176]:
df_test[df_test.pred==1].repaid.mean()

0.9216071930317505

In [177]:
df_test.pred.value_counts()

1    3559
0    2382
Name: pred, dtype: int64

In [178]:
df_test.annual_return.mean()

0.020133622934066387

##### LogisticRegression

In [179]:
from sklearn.linear_model import LogisticRegression

#lr = LogisticRegression(C=1E6, class_weight='balanced')
lr = LogisticRegression(C=1E6, class_weight={1:0.1, 0:0.9})

lr = lr.fit(d_train, l_train)
print(lr.score(d_test, l_test))
pred = lr.predict(d_test)
print(metrics.classification_report(l_test, pred))
get_pred_return(df_test, pred)

0.5078269651573809
              precision    recall  f1-score   support

           0       0.16      0.76      0.27       697
           1       0.94      0.47      0.63      5244

    accuracy                           0.51      5941
   macro avg       0.55      0.62      0.45      5941
weighted avg       0.85      0.51      0.59      5941

0:36month, 1:60month
 3    2656
Name: loan_term_year, dtype: int64
Repaid Rate:
loan_term_year
3    0.936747
Name: repaid, dtype: float64
36 months loan:2656, Annual return:0.0277
60 months loan:0, Annual return:0
predicted: (0.027681980223061675, 0)
36 months loan:5941, Annual return:0.0302
60 months loan:0, Annual return:0
all: (0.03015380427611003, 0)


In [180]:
df_test['pred'] = pred
portfolio_36 = df_test[df_test.pred==1]
get_portfolio_annual_return(portfolio_36)

36 months loan:2656, Annual return:0.0277
60 months loan:0, Annual return:0


(0.027681980223061675, 0)

In [181]:
portfolio_36.repaid.mean()

0.9367469879518072

In [182]:
get_portfolio_annual_return(df_test[df_test.repaid==0])

36 months loan:697, Annual return:-0.1523
60 months loan:0, Annual return:0


(-0.15228462368940754, 0)

In [183]:
get_portfolio_annual_return(df_test[df_test.repaid==1])

36 months loan:5244, Annual return:0.0493
60 months loan:0, Annual return:0


(0.049330908861278555, 0)

In [184]:
get_portfolio_annual_return(portfolio_36[portfolio_36.repaid==1])

36 months loan:2488, Annual return:0.0373
60 months loan:0, Annual return:0


(0.03727023563500764, 0)

In [185]:
df_test[df_test.grade=='A'].repaid.mean()

0.9434063000533903

In [186]:
get_portfolio_annual_return(df_test[df_test.grade=='A'])

36 months loan:1873, Annual return:0.0243
60 months loan:0, Annual return:0


(0.024271043992253327, 0)

In [187]:
get_portfolio_annual_return(portfolio_36[portfolio_36.grade=='D'])

36 months loan:15, Annual return:0.0575
60 months loan:0, Annual return:0


(0.05748745664970545, 0)

In [188]:
get_portfolio_annual_return(df_test[df_test.grade=='B'])

36 months loan:1944, Annual return:0.0288
60 months loan:0, Annual return:0


(0.02877328229116305, 0)

In [189]:
get_portfolio_annual_return(df_test[df_test.grade=='C'])

36 months loan:1167, Annual return:0.0328
60 months loan:0, Annual return:0


(0.0328082929862199, 0)

In [190]:
get_portfolio_annual_return(df_test[df_test.grade=='D'])

36 months loan:709, Annual return:0.0408
60 months loan:0, Annual return:0


(0.0407898769109325, 0)

In [191]:
df_test.grade.value_counts()

B    1944
A    1873
C    1167
D     709
E     184
F      52
G      12
Name: grade, dtype: int64

In [192]:
portfolio_36.grade.value_counts()

A    1803
B     783
C      51
D      15
E       3
F       1
Name: grade, dtype: int64

In [193]:
get_portfolio_annual_return(portfolio_36[portfolio_36.repaid==0])

36 months loan:168, Annual return:-0.1547
60 months loan:0, Annual return:0


(-0.15468085871308657, 0)

In [194]:
#def filter_portfolio_bi_condition(df):

In [195]:
df_test.groupby('grade').int_rate.mean()

grade
A     7.342050
B    10.993709
C    13.480300
D    15.376883
E    16.953424
F    18.540000
G    20.458333
Name: int_rate, dtype: float64

In [196]:
df_test.head()

funded_amnt  loan_term_year  int_rate grade emp_length home_ownership  \
6776          7000               3     19.39     E    7 years           RENT   
16674        21600               3     11.48     B  10+ years            OWN   
16399        11000               3     15.33     D  10+ years       MORTGAGE   
15365        11000               3     13.11     C    4 years           RENT   
10714         6875               3      6.54     A   < 1 year       MORTGAGE   

       annual_inc verification_status             purpose addr_state    dti  \
6776      43000.0            Verified  debt_consolidation         NY   5.64   
16674    138000.0        Not Verified  debt_consolidation         FL  22.09   
16399     70000.0        Not Verified  debt_consolidation         FL  24.12   
15365     50985.0        Not Verified  debt_consolidation         NJ  16.19   
10714     95000.0        Not Verified      small_business         NY   1.11   

       delinq_2yrs  inq_last_6mths  mths_since_last_delinq  \
6776           0.0             1.0                    24.0   
16674          0.0             1.0                     NaN   
16399          0.0             2.0                     NaN   
15365          0.0             1.0                     NaN   
10714          0.0             2.0                     NaN   

       mths_since_last_record  pub_rec  revol_bal  revol_util loan_status  \
6776                      NaN      0.0       5981        92.0  Fully Paid   
16674                     NaN      0.0        682         3.3  Fully Paid   
16399                     NaN      0.0      18107        68.8  Fully Paid   
15365                     NaN      0.0       3621        34.8  Fully Paid   
10714                     NaN      0.0       3640         7.4  Fully Paid   

       repaid  annual_return   total_pymnt  recoveries  pub_rec_bankruptcies  \
6776        1       0.098819   9287.014982         0.0                   0.0   
16674       1       0.003182  21806.830000         0.0                   0.0   
16399       1       0.078312  13791.968178         0.0                   0.0   
15365       1       0.065433  13303.663155         0.0                   0.0   
10714       1       0.018385   7261.198496         0.0                   0.0   

       delinq_2yrs_bi  inq_last_6mths_bi  mths_since_last_record_bi  \
6776                0                  1                          0   
16674               0                  1                          0   
16399               0                  1                          0   
15365               0                  1                          0   
10714               0                  1                          0   

       pub_rec_bi  pub_rec_bankruptcies_bi  term_code  grade_code  \
6776            0                        0          0           4   
16674           0                        0          0           1   
16399           0                        0          0           3   
15365           0                        0          0           2   
10714           0                        0          0           0   

       emp_length_code  verification_status_code  purpose_code  \
6776               7.0                         2             2   
16674             10.0                         0             2   
16399             10.0                         0             2   
15365              4.0                         0             2   
10714              0.0                         0            11   

       home_ownership_code  addr_state_code  pred  
6776                     4                4     0  
16674                    3                1     1  
16399                    0                1     0  
15365                    4                3     0  
10714                    0                4     1

In [197]:
df_test[df_test.grade=='A'][['grade', 'repaid','funded_amnt', 'total_pymnt', 'recoveries','annual_return']].head()

grade  repaid  funded_amnt   total_pymnt  recoveries  annual_return
10714     A       1         6875   7261.198496        0.00       0.018385
13046     A       1        14000  15680.190629        0.00       0.038503
13747     A       1         9800  11049.110326        0.00       0.040800
7851      A       0         6725   3350.910000      236.75      -0.188965
18320     A       1         6000   6672.905682        0.00       0.036067

### Task 3 Summary

When use class_weight to get high positive precision, majority of loans are grade A, which has low interest rate.

## Task 4: Conclusion (20 points)

Summarize:
- Your findings. 
- The limitations of your analysis.
- Possible ways to improve your analysis.
- Any thing else worth mentioning.